agent.py

```diff
@@ -989,21 +989,32 @@ def findSugarMetabolism(self):
         return max(0, self.sugarMetabolism + self.sugarMetabolismModifier)
 
     def findTribe(self):
-        if self.tags == None:
+        config = self.cell.environment.sugarscape.configuration
+        numTribes = config["environmentMaxTribes"]
+        if self.tags == None or numTribes <= 0:
             return None
-        sugarscape = self.cell.environment.sugarscape
-        numTribes = sugarscape.configuration["environmentMaxTribes"]
-        zeroes = 0
-        tribeCutoff = math.floor(len(self.tags) / numTribes)
-        for tag in self.tags:
-            if tag == 0:
-                zeroes += 1
-        self.tagZeroes = zeroes
+        self.tagZeroes = self.tags.count(0)
         for i in range(numTribes):
-            if zeroes < (i * tribeCutoff) + 1 or i == numTribes:
+            tribeSize = (len(self.tags) + 1) / numTribes
+            minZeroes = math.floor(i * tribeSize)
+            maxZeroes = math.floor((i + 1) * tribeSize) - 1
+            maxZeroes = min(maxZeroes, len(self.tags))
+            if self.tagZeroes >= minZeroes and self.tagZeroes <= maxZeroes:
                 return i
         # Default tribe
         return 0
+        '''
```

Comment:
Reverted these changes because I was not convinced they were working correctly. Please refactor in what works and cut what doesn't.

Reply:
<img width="579" alt="Screenshot 2024-06-23 at 11 42 53 PM" src="https://github.com/nkremerh/sugarscape/assets/118491522/13ebc606-f28d-41a1-bd51-c0fcecdcfd95">

It's actually _too_ correct. Randomly generating agent tags creates a binomial distribution for the number of zeroes, so the middle (index) tribes are bound to have way more agents than the outer tribes, even though the intervals are correct. Despite using random tags and discussing 3-tribe scenarios in the footnote of p. 71, GAS never mentions this. I'm pretty sure that even if we start with a uniform distribution, if cultural transmission is turned on, we will end up with a binomial distribution of tag zeroes just by the nature of the rule. We could decide tribes some other way, but any other method would interact with our transmission logic differently. Thoughts?

Reply:
This sounds like a design flaw with using binary strings as a data structure for cultural tribes. Let's keep it for now and do what we can to make it as fair as possible.